juju/model.py

```diff
@@ -0,0 +1,166 @@
+
+class Application:
+    def __init__(self, name, units):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+        self.units = units
+
+    def __hash__(self):
+        # An Application instance from one relation will not be identical to one from another relation,
+        # since they may have different views of the units of that app. However, they should be considered
+        # equivalent, since they are views of the same entity. Generally, this should only matter for using
+        # model.app as an index into a given relation's relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Application):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class Unit:
+    def __init__(self, name):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class RelationMap(dict):
+    """Map of relation names to lists of Relation instances.
+
+    The data for each relation name may be lazy-loaded.
+    """
+    def __init__(self, data):
+        super().__init__()
+        for relation_name, relations in data.items():
+            if relations is not None:
+                relations = [Relation(relation_id, relation_data)
+                             for relation_id, relation_data in sorted(relations.items())]
+            self[relation_name] = relations
+
+    def __getitem__(self, relation_name):
+        value = super().__getitem__(relation_name)
+        if value is None:
+            # Lazy-load the list of relations for this relation name.
+            output = run(['relation-ids', relation_name, '--format=yaml'], stdout=PIPE, check=True).stdout
```

Comment:
Curious choice of API by upstream developers... I'd expect stdout to be a pipe given the syntax.

Reply:
`PIPE` in this context means "pipe back to this process."  Newer Python versions (3.7+) have a `capture_output=True` param which makes way more sense to me.  We could also switch to the older `check_output()` function (which is now a wrapper around `run()`).

Reply:
The comment had that background in mind. The point was that "pipe" in pretty much every Linux API means "stream", but it's really not important for us now. In this sense, whatever works.